### DISCUSSION TITLE
Virtual BLE Devices on Windows

### DISCUSSION THREAD

**Discussion by shanedevane** [kevincar/bless#53] [Q&A]
Hi Kevin,
Newbie question - can bless create virtual bluetooth devices?
I am trying to test an android app that uses bluetooth to connect to BLE devices - and I want to create many virtual bluetooth devices to see how the android app handles it.
Can bless do that?
Thanks so much
Shane

**Comment by kevincar** [comment]
Hi Shane,

Short answer is that Bless can’t turn your computer into multiple virtual Bluetooth devices. 

Bless allows you to advertise Bluetooth Low Energy services and characteristics from your computer. So your app would discover only one device since the information would be emitted from the single Bluetooth module on your machine.

Virtualizing the hardware such that central devices would observe multiple devices is an interesting idea but Bless can’t do that. 

I’d imagine you’d need some way to control the Bluetooth hardware to emit multiple advertisements from different MAC addresses. Not sure whether that’s possible. Perhaps using bless in multiple virtual environments through Docker, virtual box, or VMWare may work, but I guess that that would depend on how they virtualize and share the same Bluetooth module on your machine.

sorry can’t be of much help here but I’m curious to know if you do find a solution.

**Reply by shanedevane** [reply]
Thanks for the reply Kevin and the idea re: VM.  That might be an interesting option.
If I find anything interesting I'll let you know.
Thanks
Shane

**Comment by unicorn200066** [comment]
Hi Shane.

I wanted to make one virtual BLE device by win10 and bless, however, I run the server and gattserver in the example folder and I found nothing by nRFConnect app. I wonder if I miss something, could you please give me some help, you can see the details in discussion #139.

Thank you so much！
yida wang
